### PR TITLE
Introduce Cyclic Quotient Singularities

### PR DESCRIPTION
**Pull request by lkastner**
Cyclic quotient singularities are two-dimensional affine normal toric varieties. Their algebraic geometry is especially well understood, and many properties can be rephrased in the associated continued fractions. They serve as an important testcase for any new translation of algebraic geometry into combinatorics via toric geometry. This PR adds the following:
- New struct wrapping the corresponding polymake object
- Extraction of the associated continued fractions
- Conversion between continued fractions and rational numbers

I don't know how self-contained the docs need to be. I tried to add the general idea and added the two main papers from deformation theory, since I use their notation.

### CODE REVIEW

**Comment by thofma** [comment]
My comments are only about the rational <-> continued fraction things.  We have some continued fraction stuff from flint in Nemo (so already in Oscar), which also handles numbers/continued fractions of arbitrary size (so not necessarily fitting into `Vector{Int}`).  It might be a good idea to expose only one version to the user, probably using a dedicated type for continued fractions. I will let @tthsqe12 comment on this, since I know that he looked at this some time ago (but I forgot the outcome/status).

**Comment by lkastner** [comment]
Thank you for the hint. I found the functionality, now I just need to know how to switch between conventions. Let me demonstrate by example:
```
julia> f = fmpq(7,5)
7//5

julia> continued_fraction(f)
3-element Vector{fmpz}:
 1
 2
 2

julia> rational_2_continued_fraction(Rational(7,5))
3-element Vector{Int64}:
 2
 2
 3

julia> 1+1/(2+Rational(1,2))
7//5

julia> 2-1/(2-Rational(1,3))
7//5
```
The first variant `1 2 2` is the Nemo variant. This corresponds to taking the `floor` of a rational number. I guess there is some switch hidden somewhere to turn it into `ceil` such that it becomes the `2 2 3`?

**Comment by thofma** [comment]
Using the `ceil` instead of `floor` is an interesting convention. I will let @tthsqe12 comment on this.

**Comment by tthsqe12** [comment]
Hehe, continued_fraction (7/5) must return [1,2,2] or the poor user will be confounded by our system. Now, the negative continued fractions are a poor default choice because the bit size of the plain vector output is exponential in the bit size of the input, which is not the case with the standard choice. You can add negative_continued_fraction(x) or continued_fraction(x, :negative) or whatever by inserting a bunch of 2's in the standard continued_fraction(x) if you want; as long as continued_fraction(QQ(7/5)) returns [1,2,2], I along with the standard user will be happy.

**Comment by thofma** [comment]
If we go down the `continued_fraction(x, :negative)` way, then we cannot just return `Vector`. Because then it becomes impossible to reconstruct the rational number. The caller of `continued_fraction_to_rational_number(::Vector)` does not necessarily know with which convention `continued_fraction` was called. Ergo every function returning a `continued_fraction` would need the keyword argument. Or, the other option would be a `ContinuedFraction` type, which can behave like a `Vector` and for which we then could have `QQ(::ContinuedFraction)` or `rational_number(::ContinuedFraction)`. Any preference?

**Comment by tthsqe12** [comment]
Sure, if `continued_fraction(7/5, :negative)` returns `[2,2,3]`, then this vector needs a proper interpretation wrapper (type stable? I don't think so). However, while negative continued fractions are non-standard, it is even more non-standard to use all positive numbers for the terms in the representation of the continued fraction. We could stick with just returning a vector (and avoid moar types) if `continued_fraction(7/5, :negative)` returns `[2,-2,3]`. Just my 2 cents.

**Comment by lkastner** [comment]
There is a mathematical interpretation to this, namely if we take the cone generated by the rays `[1,0]` and `[-q,n]`, then the continued fraction encodes the Hilbert basis. I'll do an example in polymake:
```
fulton > $cqs = new CyclicQuotient(N=>7, Q=>5);

fulton > print primitive($cqs->cone(0)->RAYS);
1 0
-5 7

fulton > print $cqs->CONTINUED_FRACTION;
2 2 3
fulton > print $cqs->cone(0)->HILBERT_BASIS;
-5 7
-2 3
-1 2
0 1
1 0
```
Please reverse the order of the Hilbert basis, denote it as `b_0=[1,0]...b_4=[-5,7]`, let `c=[c_1...c_3]` be the continued fraction, then you have `b^{i-1}+b^{i+1}=c_i*b^i`. E.g. `[1,0]+[-1,2] = 2*[0,1]` and `[-1,2]+[-5,7]=3*[-2,3]`. This is just to show that this "poor" choice is not to meant to annoy. ;)

From the continued fraction one can obtain the entire singularity and deformation theory of the associated toric variety.

But with your input I think I can wrap this in a satisfying way. Changing the signs of the `c_i` seems to be a sensible and viable choice. I will just need to document this properly. Thank you.

**Comment by tthsqe12** [comment]
This is exactly what I though this negative continued fraction would be: taylored to a specific use case. Here is my current understanding of the issue:
1. `continued_fraction(7/5)` should return something by default that at least looks like `[1,2,2]` for our users' sanity.
2. The negative continued fractions are in general too long - `(n+1)/n` has a representation of length `n` - and thus cannot be the default.
3. If we are to support `continued_fraction(7/5)` and `continued_fraction(7/5, type=:negative)`, then the latter has to return `[2,-2,3]`, otherwise we need a type-unstable wrapper of the partial quotients. OR, we change the function name...

**Comment by lkastner** [comment]
I continued reading on the topic. It is interesting that with your convention the continued fractions are so well-behaved in size. This means using `floor` and `+` has a clear advantage computation wise over the negative variant. In some literature "our" variant is referred to as "Hirzebruch-Jung continued fraction", so I am considering just renaming our functions. Even though the name becomes awkward, this is an expert setting anyway, so it should be ok.

**Comment by thofma** [comment]
Good idea. Maybe `continued_fraction_hirzebruch_jung` for better discoverability?

**Comment by tthsqe12** [comment]
<details>
  <summary>click to see a conversion function!</summary>
  
  ```julia
   # normal cf
   function cf(x::Rational{Int})
      z = Int[]
      n = numerator(x)
      d = denominator(x)
      while !iszero(d)
         q, r = divrem(n, d, RoundDown)
         push!(z, q)
         (n, d) = (d, r)
      end
      return z
   end

   # negative stuff
   function ncf(x::Rational{Int})
      z = Int[]
      n = numerator(x)
      d = denominator(x)
      while !iszero(d)
         q, r = divrem(n, d, RoundUp)
         push!(z, q)
         (n, d) = (d, -r)
      end
      return z
   end

   # convert from normal to negative
   function cf2ncf(a::Vector{T}) where T
      z = T[]
      n = length(a)
      for i in 1:n
         ai = a[i]
         if iseven(i)
            ai < 2^30 || @warn "blowing up your memory"
            while (ai -= 1) > 0
               push!(z, T(2))
            end
         else
            push!(z, ai + (1 < i < n) + (1 < n))
         end
      end
      z
   end

   function testit(x)
      a = ncf(x)
      b = cf2ncf(cf(x))
      if a != b
         @show a
         @show b
         @assert false
      end
   end

   testit(1//1)
   testit(1 + 1//2)
   testit(1 + 1//(2 + 1//(3 + 1//(4 + 1//(5 + 1//(6 + 1//7))))))
   testit(1 + 1//(2 + 1//(3 + 1//(4 + 1//(5 + 1//(6 + 1//(7 + 1//8)))))))

   for i in 1:1000000
      testit(rand(-10000:10000)//rand(1:10000))
   end
  ```
</details>

**Comment by fingolfin** [comment]
So what's the status of this? Can it be merged?

**Comment by lkastner** [comment]
Sorry, I was distracted.

@tthsqe12 : Thank you for the conversion functions. Do I understand you correctly, that I just add them with the cyclic quotient singularities?

**Comment by tthsqe12** [comment]
I was not suggesting that you add anything. I was merely pointing out that conversion between the two is possible. While the for loop around the divrem is an easy implementation, this can get slow when the input is big: if we ever wanted the negative cf for fmpqs, I would recommend calculating the normal cf via Nemo, then converting. Do you want `continued_fraction_hirzebruch_jung(::fmpq)` here?

**Comment by lkastner** [comment]
You mean in Nemo? And then change the `continued_fraction_hirzebruch_jung(CyclicQuotientSingularity)` accordingly? For me this sounds like a good solution.

File: src/ToricVarieties/CyclicQuotientSingularities.jl [hunk 1]
```diff
@@ -0,0 +1,177 @@
...
+ 37
+
+julia> continued_fraction_hirzebruch_jung_to_rational(cf)
+2464144958//145732115
+
+julia> r == continued_fraction_hirzebruch_jung_to_rational(cf)
+true
+```
+"""
+function rational_to_continued_fraction_hirzebruch_jung(r::Rational)
```
Comment (tthsqe12) [inline src/ToricVarieties/CyclicQuotientSingularities.jl]: I see that the return is a vector of machine integers, but the partial quotients can't be guaranteed to fit in machine integers unless the input is also machine-sized. Maybe `r::Rational{Int}` here?

**Comment by tthsqe12** [comment]
Sorry, for the confusion, but I am not suggesting putting `continued_fraction_hirzebruch_jung` into nemo. I am saying that *if* it is desired to have it operating on the oscar multiprecision rational type `fmpq`, it would be best implemented as
`rational_to_continued_fraction_hirzebruch_jung(r::fmpq) = cf2nfc(continued_fraction(r))` where cf2nfc is the conversion function. Basically, everything I said can be ignored (except the comment on the machine-sized output on non-machine-sized input I just made).

**Comment by tthsqe12** [comment]
Ok, this looks fine: There is a `CyclicQuotientSingularity(n::fmpz, q::fmpz)` which takes two arbitrary precision integers. You can ask for various continued fractions on this objects and they return vectors of fmpzs. There is also a pair of functions for converting between fmpqs and these negative continued fractions, again operating on all multiprecision types.

**Pull request merged**